### DISCUSSION TITLE
Pyright complains when a f-string is passed into a custom alias of `typing.Annotated`

### DISCUSSION THREAD

**Discussion by EtaoinWu** [microsoft/pyright#6713] [Q&A]
With the following code:

```python
from typing import Annotated, reveal_type
MyAnn = Annotated

def f1(n):
    # this typechecks
    def g(x: Annotated[list[int], f"has a length of {n}"]) -> int:
        return sum(x)
    return g

def f2(n):
    # pyright complains `annotationFormatString`
    def g(x: MyAnn[list[int], f"has a length of {n}"]) -> int:
        reveal_type(x)
        return sum(x)
    return g

l = [1, 1, 4, 5, 1, 4]
print(f1(6)(l))
print(f2(6)(l))
```

Pyright complains at `f2`'s inner function `g`, because "Type annotations cannot use format string literals (f-strings)". However, the type for `f2.g.x` is inferred correctly as `list[int]`, revealing that pyright understands `MyAnn`. 

Furthermore, the fact that f-strings are specifically treated here doesn't make a lot of sense to me; there _can_ be other runtime-calculated values like:

```python
from dataclasses import dataclass

@dataclass
class HasALengthOf:
    length: int

def f3(n):
    def g(x: Annotated[list[int], HasALengthOf(n)]) -> int:
        return sum(x)
    return g

def f4(n):
    # this is correct, even though HasALengthOf(n) is computed at run-time
    def g(x: MyAnn[list[int], HasALengthOf(n)]) -> int:
        return sum(x)
    return g
```

Might raise this as an issue?

**Comment by EtaoinWu** [comment]
With a bit of digging into the internals, I realized that `annotationFormatString` is reported during parsing, which happens before the binding pass, and the parser have no idea that `MyAnn` is an alias of `Annotated`.

**Reply by erictraut** [reply]
Yes, don't use custom aliases of `Annotated`. It's a special form, and you should use the version exported from `typing` or `typing_extensions`.

**Reply by erictraut** [reply]
If you want to use a different name for `Annotated` locally, pyright does support the `as` form of import, as in `from typing import Annotated as MyAnn`. But pyright doesn't support re-exporting `MyAnn` and then importing it and using it in another module.

**Comment by EtaoinWu** [comment]
The code I showed above is a minimal example. Here is the X problem:

I encounter this issue when using [jaxtyping](https://github.com/google/jaxtyping). In short, this package supports type annotation and runtime typechecking of ndarrays. As an example, it allows you to write `x: Float[np.ndarray, "n 4"]`, which says that `x` is an `np.ndarray` with a float-like dtype and a shape of `(n, 4)` where `n` is an arbitrary number.

During static type checking (if `TYPE_CHECKING` is true), `Float` is [defined as](https://github.com/google/jaxtyping/blob/7df267efa4986f4ea7c40f6e5177e30f95a1d0cd/jaxtyping/_indirection.py#L29) an (re-exported) alias of `Annotated`:

```python
# somewhere in jaxtyping's internal
from typing import (
    Annotated as Float,
    # similarly for other dtypes like Bool, UInt32, ...
)
```

and therefore a static typechecker is able to infer `x` to be just an `np.ndarray`, ignoring the shape and dtype information. The package also provides facilities to check the dtype and shape of arrays at runtime, but that's irrelevant here.

Because python function annotations are evaluated at function creation time, this allows us to write (proposed by [jaxtyping#93](https://github.com/google/jaxtyping/issues/93)):

```python
import numpy as np
from jaxtyping import Float

def f(n):
    # pyright complains here
    def g(x: Float[np.ndarray, f"m {n}"]) -> Float[np.ndarray, f"{n}"]:
        return x.sum(axis=0)
    return g

print(f(4)(np.ones((2, 4))))
```

> [...] pyright doesn't support re-exporting `MyAnn` and then importing it and using it in another module.

So that's the problem -- if this is *unsupported explicitly* by pyright's parser, I believe that we have to suppress or bypass the error.

I found a very dumb way of bypassing this. Since pyright allows function calls during type annotation, we can use an identity function:

```python
def identity(x):
    return x

def f(n):
    def g(x: Float[np.ndarray, identity(f"m {n}")]) -> Float[np.ndarray, identity(f"{n}")]:
        return x.sum(axis=0)
    return g
```

This also typechecks under mypy.

**Reply by erictraut** [reply]
That's a pretty gross hack. Let's me see if I can defer the error detection in pyright so it occurs in a later stage than the parser. If jaxtyping is relying on this behavior, then you're not the only person who will run into this issue. I've filed [this issue](https://github.com/microsoft/pyright/issues/6714) to track this.

**Comment by EtaoinWu** [comment]
Closed, as #6714 is fixed by #6719! That's an impressive speed.
